src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java

```diff
@@ -0,0 +1,203 @@
+import com.hivemq.HiveMQServer;
+import com.hivemq.bootstrap.ioc.GuiceBootstrap;
+import com.hivemq.common.shutdown.HiveMQShutdownHook;
+import com.hivemq.common.shutdown.ShutdownHooks;
+import com.hivemq.configuration.ConfigurationBootstrap;
+import com.hivemq.configuration.HivemqId;
+import com.hivemq.configuration.info.SystemInformationImpl;
+import com.hivemq.configuration.service.FullConfigurationService;
+import com.hivemq.embedded.EmbeddedHiveMQ;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.extension.sdk.api.annotations.Nullable;
+import com.hivemq.persistence.PersistenceStartup;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+/**
+ * @author Georg Held
+ */
+public class EmbeddedHiveMQImpl implements EmbeddedHiveMQ {
+
+    private static final @NotNull Logger log = LoggerFactory.getLogger(EmbeddedHiveMQImpl.class);
+
+    private final @NotNull SystemInformationImpl systemInformation;
+    private final @NotNull MetricRegistry metricRegistry;
+    private @Nullable FullConfigurationService configurationService;
+    private @Nullable Injector injector;
+
+    private volatile @NotNull State currentState;
+    private volatile @Nullable CompletableFuture<Void> startFuture;
+    private volatile @Nullable CompletableFuture<Void> stopFuture;
+
+    EmbeddedHiveMQImpl(
+            final @Nullable File conf, final @Nullable File data, final @Nullable File extensions) {
+
+        currentState = State.NOT_STARTED;
+        systemInformation = new SystemInformationImpl(true, true, conf, data, extensions);
+        // we create the metric registry here to make it accessible before start
+        metricRegistry = new MetricRegistry();
+    }
+
+    private enum State {
+
+        NOT_STARTED,
+        STARTING,
+        RUNNING,
+        STOPPING,
+        STOPPED;
+
+        private boolean isHigher(final @NotNull State other) {
+            return this.ordinal() > other.ordinal();
+        }
+
+        private boolean isLower(final @NotNull State other) {
+            return this.ordinal() < other.ordinal();
+        }
+    }
+
+    private synchronized void advanceState(final @NotNull State newState) {
+        if ((currentState == State.STOPPED && newState == State.NOT_STARTED)
+                || newState.isHigher(currentState)) {
+            log.debug("Advancing EmbeddedHiveMQState from \"{}\" to \"{}\"", currentState, newState);
+            currentState = newState;
+        }
+    }
+
+    @Override
+    public synchronized @NotNull CompletableFuture<Void> start() {
+        advanceState(State.NOT_STARTED);
+        if (currentState == State.NOT_STARTED) {
+            advanceState(State.STARTING);
+            log.info("Starting EmbeddedHiveMQ.");
+
+            startFuture = new CompletableFuture<>();
+            CompletableFuture.runAsync(() -> {
```

Comment:
I don't think it is a good idea to use a shared executor for the startup, because this blocks for some time.
I think a new thread here would be the best solution

Reply:
I dont like spawning a new thread per Embedded HiveMQ. What exactly is the advantage for this? Now it is configurable for the user. ut perhapps we should make this configuration explicit? What do you think about allowing to set an executor in the builder?

Reply:
HiveMQ takes a long time to start (even when we optimize it to under 1sec this is still very long). Blocking a shared executor for such a long time seems wrong to me, we actually stall other work that is totally unrelated during the HiveMQ startup/shutdown.
I would not add a configuration for that right now in the builder, as the right way is to make the start of HiveMQ return a Future anyway.

Reply:
Done